cpp/src/arrow/compare.h

```diff
@@ -142,4 +152,208 @@ bool ARROW_EXPORT
 ScalarApproxEquals(const Scalar& left, const Scalar& right,
                    const EqualOptions& options = EqualOptions::Defaults());
 
+class OrderOptions {
+ public:
+  // Whether NaNs are considered least in order
+  bool nans_least() const { return nans_least_; }
+
+  OrderOptions nans_least(bool v) const {
+    auto res = OrderOptions(*this);
+    res.nans_least_ = v;
+    return res;
+  }
+
+  /// The absolute tolerance for approximate comparisons of floating-point values.
+  double atolf() const { return atolf_; }
+
+  /// Return a new EqualOptions object with the "atol" property changed.
+  OrderOptions atolf(double v) const {
+    auto res = OrderOptions(*this);
+    res.atolf_ = v;
+    return res;
+  }
+
+  /// The absolute tolerance for approximate comparisons of integer values.
+  uint64_t atold() const { return atold_; }
+
+  /// Return a new EqualOptions object with the "atol" property changed.
+  OrderOptions atold(double v) const {
+    auto res = OrderOptions(*this);
+    res.atold_ = v;
+    return res;
+  }
+
+  static OrderOptions Defaults() { return {}; }
+
+ protected:
+  double atolf_ = kDefaultOrderFloatingPointAbsoluteTolerance;
+  uint64_t atold_ = kDefaultOrderIntegerAbsoluteTolerance;
+  bool nans_least_ = false;
+};
+
+namespace internal {
+
+template <typename T>
+T TypeAdjustedTolerance(double tolerance) {
+  T max = std::numeric_limits<T>::max();
+  return tolerance >= max ? max : static_cast<T>(tolerance);
+}
+
+template <typename T>
+T TypeAdjustedTolerance(uint64_t tolerance) {
+  T max = std::numeric_limits<T>::max();
+  return tolerance >= static_cast<uint64_t>(max) ? max : static_cast<T>(tolerance);
+}
+
+template <typename T, template <typename> class Op,
+          typename = std::enable_if<std::is_base_of<Scalar, T>::value>>
+bool NumericScalarCompare(const T& left, const T& right, const OrderOptions& options) {
```

Comment:
I'm not sure why we're exposing these publicly? Aren't the high-level variants taking a `const Scalar&` sufficient?